Issue#10801 : Made limits work with a binomial expression

rehassachdeva opened the pull request:
This patch fixes #10801 .

The issue was that this doesn't work:
`limit(16**k/(k*binomial(2*k,k)**2), k, oo)`
In general limit doesn't work with such expressions. 
So to fix it we can check if it contains binomial and then convert it to factorial form.
Now it works as shown:

```
In [1]: limit(16**k/(k*binomial(2*k,k)**2), k, oo)
Out[1]: π
```

Please review.

rehassachdeva committed:
series: Made limit work with binomial() and added test

Made limits work with an expression containing binomial by converting
the expression to factorial form before evaluating it. This should not be
done for expressions which contain infinite sequences. Accordingly the
check was added.

Example:

In [1]: limit(16**k/(k*binomial(2*k,k)**2), k, oo)
Out[1]: π

leosartaj commented:
You should use a more informative PR title. Please read [this](https://github.com/sympy/sympy/wiki/Development-workflow#writing-pull-request-title-and-description).

leosartaj commented on sympy/series/limits.py:
```diff
@@ -136,6 +136,11 @@ def doit(self, **hints):
             return e
 
+        # limit can work with binomial by converting to factorial first
+        if (e.has(binomial) and not (hints.get('sequence', True)
```
Just ``if e.has(binomial)`` would be better. I would prefer ``limit_seq`` to be a last resort.

leosartaj replied:
I see that the ``limit(binomial(2*x, x) / Sum(binomial(2*y, y), (y, 1, x)), x, oo) == S(3) / 4`` gets stuck, if only the condition for binomial is checked.

rehassachdeva replied:
Hi. If this check isn't included then this fails.
`assert (limit(binomial(2*x, x) / Sum(binomial(2*y, y), (y, 1, x)), x, oo) ==  S(3) / 4)`

leosartaj replied:
This appears a little hacky. I would go with implementing ``_eval_as_tractable``.

```
>>> e = 16**k / (k * binomial(2*k, k)**2)
>>> limit(e.rewrite(gamma), k, oo) # works
pi
```
By applying the patch,

```
diff --git a/sympy/functions/combinatorial/factorials.py b/sympy/functions/combinatorial/factorials.py
index 48ea175..0b7adb2 100644
--- a/sympy/functions/combinatorial/factorials.py
+++ b/sympy/functions/combinatorial/factorials.py
@@ -853,6 +853,8 @@ def _eval_rewrite_as_gamma(self, n, k):
         from sympy import gamma
         return gamma(n + 1)/(gamma(k + 1)*gamma(n - k + 1))
 
+    _eval_rewrite_as_tractable = _eval_rewrite_as_gamma
+
     def _eval_rewrite_as_FallingFactorial(self, n, k):
         if k.is_integer:
             return ff(n, k) / factorial(k)
```

```
>>> limit(e.rewrite(gamma), k, oo)
pi
>>> limit(e, k, oo) # This seems odd
oo
```

ping @asmeurer

rehassachdeva replied:
Hi. Could you please guide how `_eval_as_tractable` should work?
I tried implementing `_eval_as_tractable` . But i was confused. For trying, I simply made it return factorial form, same as  `_eval_as_factorial`. and calling e.rewrite('tractable') in limits.py. But this was unable to evaluate the limit and returned the limit expr as such.

leosartaj replied:
I tried something similar. As you said, returning in ``factorial`` form did not help, so I tried it with ``gamma``, which appears to give wrong result. You do need not to call it anywhere. It is already called [here](https://github.com/sympy/sympy/tree/master/sympy/series/gruntz.py#L411). Also, You need to implement ``_eval_rewrite_as_tractable`` and not ``eval_as_tractable``.

leosartaj replied:
I tried something similar. As you said, returning in ``factorial`` form did not help, so I tried it with ``gamma``, which appears to give wrong result. You do need not to call it anywhere. It is already called [here](https://github.com/sympy/sympy/tree/master/sympy/series/gruntz.py#L411). Also, You need to implement ``_eval_rewrite_as_tractable`` and not ``eval_as_tractable``.

rehassachdeva replied:
yes i tried with name `_eval_rewrite_as_tractable'. And i was grepping for "_rewrite(tractable" as with "_rewrite(factorial" no quotes. that's why i didn't follow that. But i tried with 'e.rewrite('tractable')'. 
And i was unable to reproduce the error with gamma
```
In [9]: e = 16**k / (k * binomial(2*k, k)**2)

In [10]: limit(e.rewrite(gamma), k, oo) # works
Out[10]: π

In [11]: limit(e.rewrite(gamma), k, oo)
Out[11]: π

In [12]: limit(e, k, oo) # This seems okay
Out[12]: π
```

leosartaj replied:
After applying, on the current **master**

```
diff --git a/sympy/functions/combinatorial/factorials.py b/sympy/functions/combinatorial/factorials.py
index 48ea175..0b7adb2 100644
--- a/sympy/functions/combinatorial/factorials.py
+++ b/sympy/functions/combinatorial/factorials.py
@@ -853,6 +853,8 @@ def _eval_rewrite_as_gamma(self, n, k):
         from sympy import gamma
         return gamma(n + 1)/(gamma(k + 1)*gamma(n - k + 1))
 
+    _eval_rewrite_as_tractable = _eval_rewrite_as_gamma
+
     def _eval_rewrite_as_FallingFactorial(self, n, k):
         if k.is_integer:
             return ff(n, k) / factorial(k)
```

I still get,

```
>>> e = 16**k / (k * binomial(2*k, k)**2)
>>> limit(e.rewrite(gamma), k, oo)
pi
>>> limit(e.rewrite(factorial), k, oo)
pi
>>> limit(e, k, oo)
oo
```

rehassachdeva replied:
This is strange.

rehassachdeva replied:
Sorry, I didn't see the patch you applied, correctly. 
Yes i'm getting the same errors if we implement `_eval_rewrite_as_tractable` to return gamma (oo) and to return factorial (doesn't evaluate) . 
So what should be done?

jksuom replied:
Did you try what ``gamma._eval_rewrite_as_tractable`` does?

leosartaj replied:
Something like does seem to work.

```
diff --git a/sympy/functions/combinatorial/factorials.py b/sympy/functions/combinatorial/factorials.py
index 48ea175..fc11ad0 100644
--- a/sympy/functions/combinatorial/factorials.py
+++ b/sympy/functions/combinatorial/factorials.py
@@ -853,6 +853,9 @@ def _eval_rewrite_as_gamma(self, n, k):
         from sympy import gamma
         return gamma(n + 1)/(gamma(k + 1)*gamma(n - k + 1))
 
+    def _eval_rewrite_as_tractable(self, n, k):
+        return self._eval_rewrite_as_gamma(n, k).rewrite('tractable')
+
     def _eval_rewrite_as_FallingFactorial(self, n, k):
         if k.is_integer:
             return ff(n, k) / factorial(k)
```

```
>>> e = 16**k / (k * binomial(2*k, k)**2)
>>> limit(e, k, oo)
pi
```

Though, I am still worried as to why does, setting ``_eval_rewrite_as_tractable`` equal to ``_eval_as_rewrite_gamma`` does return ``oo``. A bug in the implementation of gruntz perhaps?

jksuom replied:
If I recall correctly, rewrite(tractable) is called only once in gruntz. If that rewrites binomial in terms of gamma, that will not be rewritten as tractable any more.

leosartaj commented on sympy/series/tests/test_limits.py:
```diff
@@ -456,6 +456,11 @@ def test_issue_8730():
 
+def test_issue_10801():
+    # make sure limits work with binomial
+    assert limit(16**k/(k*binomial(2*k,k)**2), k, oo) == pi
```
``limit(16**k/(k*binomial(2*k,k)**2), k, oo)`` -> ``limit(16**k / (k * binomial(2*k, k)**2), k, oo)``

rehassachdeva replied:
Code style fix noted.
The suggestion for using `gamma._eval_rewrite_as_tractable` is essentially 
```
In [1]: e = 16**k / (k * binomial(2*k, k)**2)

In [2]: limit(e.rewrite(gamma).rewrite('tractable'), k, oo)
Out[2]: π
```
and it works. 
This alone also works:
```
In [3]: limit(e.rewrite(gamma), k, oo)
Out[3]: π
```
Then why does simply ` _eval_rewrite_as_tractable = _eval_rewrite_as_gamma` not work?

leosartaj replied:
Actually [here](https://github.com/sympy/sympy/tree/master/sympy/series/gruntz.py#L411) the expression is converted into tractable form. In the case ``_eval_rewrite_as_tractable = _eval_rewrite_as_gamma``, ``binomial`` is just converted into ``gamma`` form. The current implementation wants  ``gamma`` to also be converted into tractable form.

rehassachdeva replied:
Yes that's exactly what i meant.
so should we go ahead with the 
```
def _eval_rewrite_as_tractable(self, n, k):
    return self._eval_rewrite_as_gamma(n, k).rewrite('tractable')
```
approach?

leosartaj replied:
I'm +1 towards this fix.

rehassachdeva committed:
series: Improved hacky solution for making limit work with binomial()

This patch implements the method _eval_rewrite_as_tractable for binomial
class. This makes limits work with an expression containing binomial by
converting them to gamma form which is subsequently rewritten as tractable.
Hacky conditional check for expressions containing infinite sequences is
now not needed.
This patch also fixes the code style by adding spaces around binary operators
in the test added for this issue.

rehassachdeva commented:
Have updated. Please review. 
Thanks

leosartaj commented:
+1 to merge, if tests pass.

leosartaj commented:
Different Fix for the same issue #10825.

jksuom commented:
I'd prefer this one.

leosartaj commented:
Thanks! @rehassachdeva
